File: python/react/react.py
class InputCell:
    def __init__(self, initial_value):
        self._value = initial_value
        self._dependents = set() # Stores ComputeCells that depend on this cell

    @property
    def value(self):
        return self._value

    @value.setter
    def value(self, new_value):
        if new_value == self._value:
            return # No change, do nothing

        self._value = new_value

        # --- Update propagation ---
        # Keep track of cells that changed during this update cycle
        # Store the value they had *before* this update cycle started
        changed_in_cycle = {}

        # Use a queue for processing, allowing re-queueing
        queue = list(self._dependents)
        # Keep track of cells added to the queue to avoid redundant immediate adds,
        # but allow re-adding later if an input changes. Using a set for quick lookups.
        queued_or_processed = set(self._dependents)

        idx = 0
        while idx < len(queue): # Process queue elements without popping until done
            cell = queue[idx]
            idx += 1 # Move to next item for next iteration

            # Need to handle potential AttributeError if cell is not ComputeCell
            # Although dependents should always be ComputeCells based on _add_dependent usage
            if not isinstance(cell, ComputeCell):
                 continue # Should not happen in valid usage

            old_value = cell.value # Get value before recomputing
            new_value = cell._compute() # Recompute based on current inputs

            if new_value != old_value:
                # Value changed, record original value if first change in cycle
                if cell not in changed_in_cycle:
                    changed_in_cycle[cell] = old_value

                # Update the cell's internal value
                cell._update_value(new_value) # Use internal update method

                # Add dependents to the queue for potential updates
                for dep in cell._dependents:
                    # Add dependent to queue if it hasn't been queued/processed *yet* in this wave
                    # OR if it has already been processed but needs re-evaluation because this input changed.
                    # Simple approach: always add dependents if the current cell changed.
                    # Let the `if new_value != old_value` check prevent infinite loops for stable values.
                    # We might process cells multiple times, but it ensures correctness.
                    if dep not in queued_or_processed:
                         queue.append(dep)
                         queued_or_processed.add(dep)
                    else:
                         # If already processed, ensure it's re-added to the queue
                         # if it's not already pending later in the queue.
                         # This check avoids adding duplicates right next to each other.
                         if dep not in queue[idx:]:
                              queue.append(dep)


        # --- Callback firing ---
        # After the system has stabilized, call callbacks for compute cells whose
        # final value is different from their value before the update cycle began.
        for cell, original_value in changed_in_cycle.items():
             if isinstance(cell, ComputeCell): # Ensure it's a compute cell
                 # Check if the *final* value differs from the *original* value
                 if cell.value != original_value:
                      cell._fire_callbacks()


    def _add_dependent(self, compute_cell):
        """Internal method called by ComputeCell during its init."""
        # Ensure we only add ComputeCells as dependents
        if isinstance(compute_cell, ComputeCell):
             self._dependents.add(compute_cell)


class ComputeCell:
    def __init__(self, inputs, compute_function):
        self._inputs = inputs
        self._compute_function = compute_function
        self._callbacks = []
        self._dependents = set() # Stores ComputeCells that depend on this cell

        # Register this cell as a dependent of its inputs
        # Crucially, this must happen *before* the initial compute
        for inp in self._inputs:
            # Works for both InputCell and ComputeCell inputs
            inp._add_dependent(self)

        # Calculate and store initial value
        # This might trigger _compute on cells that haven't fully initialized dependents yet?
        # Let's ensure _add_dependent is robust.
        self._value = self._compute()


    @property
    def value(self):
        """Return the current value of the cell."""
        return self._value

    def add_callback(self, callback):
        """Add a callback function."""
        self._callbacks.append(callback)

    def remove_callback(self, callback):
        """Remove a previously added callback."""
        try:
            self._callbacks.remove(callback)
        except ValueError:
            pass # Callback not found

    def _add_dependent(self, compute_cell):
        """Internal method called by other ComputeCells during their init."""
         # Ensure we only add ComputeCells as dependents
        if isinstance(compute_cell, ComputeCell):
            self._dependents.add(compute_cell)

    def _compute(self):
        """Internal method to compute the value based on inputs."""
        # This might be called during init before all dependencies are set?
        # The value property access should be safe.
        return self._compute_function([i.value for i in self._inputs])

    def _update_value(self, new_value):
        """Internal method to update the cell's value during propagation."""
        self._value = new_value

    def _fire_callbacks(self):
        """Internal method to call callbacks with the current value."""
        for cb in self._callbacks:
            cb(self._value)

File: javascript/react/react.js
export class InputCell {
  constructor(value) {
    this._value = value;
    this._dependents = new Set(); // Cells that depend on this one
  }

  get value() {
    return this._value;
  }

  setValue(newValue) {
    if (this._value !== newValue) {
      this._value = newValue;
      this._propagate(); // Start the propagation wave
    }
  }

  _propagate() {
    const queue = Array.from(this._dependents);
    const visitedInWave = new Set(); // Track cells processed in this wave to prevent cycles/redundancy
    const changedInWave = new Set(); // Track ComputeCells whose values changed

    let count = 0;
    const MAX_COUNT = 1000; // Safety break

    // Seed the queue with initial dependents
    const initialQueue = Array.from(this._dependents);
    const processingQueue = [...initialQueue]; // Use a copy to iterate

    while (processingQueue.length > 0 && count < MAX_COUNT) {
      count++;
      const cell = processingQueue.shift();

      // Skip if already processed *successfully* in this wave
      if (visitedInWave.has(cell)) {
        continue;
      }

      let changed = false;
      if (cell instanceof ComputeCell) {
        const oldValue = cell.value;
        // Recompute based on potentially updated inputs. _computeValue reads current input values.
        const newValue = cell._computeValue();

        if (newValue !== oldValue) {
          cell._value = newValue; // Update internal value *before* adding dependents
          changedInWave.add(cell); // Mark that this cell changed
          changed = true;
        }
      } else {
          // If it's not a ComputeCell, it cannot change value based on inputs (it's an InputCell)
          // Or handle other cell types if they exist
      }

      // Mark as processed for this wave
      visitedInWave.add(cell);

      // If the cell changed OR if it's the initial input propagation, add its dependents
      // We need to ensure dependents are always considered if an upstream cell is processed.
      // Let's refine: Add dependents only if the current cell *could* influence them.
      // For ComputeCells, this happens if their value changed.
      if (changed && cell instanceof ComputeCell) {
          cell._dependents.forEach(dep => {
            // Add dependent to the queue. If it was already processed,
            // it might need reprocessing if its inputs changed.
            // Remove from visited so it can be processed again if needed.
            visitedInWave.delete(dep); // Allow reprocessing
            // Avoid adding duplicates to the queue itself
            if (!processingQueue.includes(dep)) {
                 processingQueue.push(dep);
            }
          });
      } else if (this._dependents.has(cell)) {
          // If it's a direct dependent of the initial InputCell, ensure its dependents are queued once.
          // This seems overly complex. Let's simplify the logic.

          // Simpler approach: Always add dependents if the current cell was processed.
          // The `visitedInWave` check prevents infinite loops. If a dependent needs
          // re-evaluation due to another input changing, it will be processed again.
           if (cell instanceof ComputeCell) { // Only compute cells have dependents to propagate to
               cell._dependents.forEach(dep => {
                   if (!processingQueue.includes(dep)) { // Basic check to avoid queue bloat
                       processingQueue.push(dep);
                   }
               });
           }
      }
    }


    if (count >= MAX_COUNT) {
      console.warn("Max propagation limit reached; potential cycle or large graph.");
    }

    // After propagation is complete, run callbacks *only once* for cells that changed.
    changedInWave.forEach(cell => cell._runCallbacks());
  }


  _addDependent(cell) {
    this._dependents.add(cell);
  }

  _removeDependent(cell) {
    this._dependents.delete(cell);
  }
}

export class ComputeCell {
  constructor(inputCells, fn) {
    this._inputs = inputCells;
    this._fn = fn;
    this._dependents = new Set();
    this._callbacks = new Set(); // Stores CallbackCell instances
    this._value = undefined; // Initialize value

    // Register this compute cell as a dependent of its inputs
    this._inputs.forEach(input => {
        if (!input || typeof input._addDependent !== 'function') {
             console.error("Invalid input cell provided:", input);
             throw new Error("Invalid input cell provided to ComputeCell constructor.");
        }
        input._addDependent(this);
    });


    // Calculate initial value - This should ideally happen within a propagation system
    // but for initialization, we compute directly. Callbacks are not run here.
    // Handle potential errors during initial computation
    try {
        this._value = this._computeValue();
    } catch (e) {
        console.error("Error during initial computation:", e);
        this._value = undefined; // Or some default error state
    }
  }

  get value() {
    // In this eager system, the value should be up-to-date.
    return this._value;
  }

  // Internal method to compute value based on current input cells
  _computeValue() {
     // Pass the array of input cell instances to the compute function
     try {
        // The compute function provided by the test expects the cell instances
        return this._fn(this._inputs);
     } catch (e) {
        console.error("Error computing value:", e);
        // Decide how to handle compute errors (e.g., return undefined, keep old value?)
        return this._value; // Keep old value on error for now
     }
  }

  addCallback(cbCell) {
    if (!(cbCell instanceof CallbackCell)) {
      throw new Error("Argument must be an instance of CallbackCell");
    }
    this._callbacks.add(cbCell);
  }

  removeCallback(cbCell) {
    if (!(cbCell instanceof CallbackCell)) {
      throw new Error("Argument must be an instance of CallbackCell");
    }
    this._callbacks.delete(cbCell);
  }

  // Internal method called by the propagation logic after value has changed
  _runCallbacks() {
    this._callbacks.forEach(cbCell => {
      cbCell._executeCallback(this);
    });
  }

  // Methods for graph structure
  _addDependent(cell) {
    this._dependents.add(cell);
  }

  _removeDependent(cell) {
    this._dependents.delete(cell);
  }

  // Optional: Cleanup method
  dispose() {
    this._inputs.forEach(input => {
        if (input && typeof input._removeDependent === 'function') {
            input._removeDependent(this);
        }
    });
    this._callbacks.clear();
    this._dependents.clear(); // Prevent this cell from being updated or updating others
  }
}

export class CallbackCell {
  constructor(fn) {
    if (typeof fn !== 'function') {
      throw new Error("CallbackCell requires a function argument.");
    }
    this.fn = fn; // The actual callback function
    this.values = []; // Stores the history of values reported by the callback
  }

  // Internal method called by ComputeCell to execute the callback and store the value
  _executeCallback(computeCellInstance) {
    try {
        this.values.push(this.fn(computeCellInstance));
    } catch (e) {
        console.error("Error executing callback:", e);
        // Decide how to handle callback errors
    }
  }
}
